new: arm64 support

FedeDP opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/CONTRIBUTING.md) file.
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area build

/area cmd

/area pkg

> /area docs

> /area tests


**What this PR does / why we need it**:

* Builds and pushes driverkit builder images for arm64 too
* Adds support for arm64 in various builders
* Adds support for cross-building on docker processor, using `multiarch/qemu-user-static`. By default, current running architecture will be built; but one has the option to force a different arch with the `architecture` option.
* Added option to pass `kernelurls` directly from the config file/cli argument. [kernel-crawler](https://github.com/falcosecurity/kernel-crawler) will dump driverkit-like configs with the kernel urls already provided, as it fetched them.
* Bumped all the deps versions

 Still needed:

* ~~Double check the very first commit; i am not sure if CircleCI will work. If anyone with a strong knowledge could double check it, it would really help me!~~ Fixed with 6f9deb0f27809207aed3c35787dc1400a8f2d776

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #142

supersedes #119 

**Special notes for your reviewer**:

NOTE: the diff is big but most of the changes are from go.sum.

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
new: added support for arm64 builders
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/driverkit/pull/143#" title="Author self-approved">FedeDP</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fntlnz** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fntlnz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fdriverkit).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/driverkit/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fntlnz"]} -->

FedeDP commented:
/cc @zuc @maxgio92

FedeDP commented:
I am not getting why tests are failing. On my setup they pass with no issues.
Log does not tell much: https://app.circleci.com/pipelines/github/falcosecurity/driverkit/232/workflows/a32bfd26-2481-4de6-b03d-40ea06fe162e/jobs/314

FedeDP commented:
Builders:
- [ ] AmazonLinux -> after applying #141  -> build gets interrupted while building the kernel module:
```
DEBU Q+ make KERNELDIR=/tmp/kernel CC=/usr/bin/gcc-8 LD=/usr/bin/ld.bfd CROSS_COMPILE=
DEBU *make -C /tmp/kernel M=/tmp/driver modules
DEBU *make[1]: Entering directory '/tmp/kernel'
DEBU   CC [M]  /tmp/driver/main.o
DEBU -  CC [M]  /tmp/driver/dynamic_params_table.o
DEBU &  CC [M]  /tmp/driver/fillers_table.o
DEBU $  CC [M]  /tmp/driver/flags_table.o
DEBU #  CC [M]  /tmp/driver/ppm_events.o
DEBU $  CC [M]  /tmp/driver/ppm_fillers.o
DEBU log pipe close                                error=EOF
DEBU context canceled
``` 
- [ ] Ubuntu/Debian -> seems related to gcc version:
```
make KERNELDIR=/tmp/kernel-download/usr/src/linux-headers-5.11.0-49-generic 
make -C /tmp/kernel-download/usr/src/linux-headers-5.11.0-49-generic M=/tmp/driver modules 
make[1]: Entering directory '/tmp/kernel-download/usr/src/linux-headers-5.11.0-49-generic'
CC [M]  /tmp/driver/main.o   
gcc: error: unrecognized command line option '-mbranch-protection=pac-ret+leaf+bti' 
gcc: error: unrecognized command line option '-mstack-protector-guard=sysreg'; did you mean '-fstack-protector-strong'? 
gcc: error: unrecognized command line option '-mstack-protector-guard-reg=sp_el0'; did you mean '-fstack-protector-all'? 
gcc: error: unrecognized command line option '-mstack-protector-guard-offset=1344'; did you mean '-fstack-protector-strong'? 
make[2]: *** [scripts/Makefile.build:288: /tmp/driver/main.o] Error 1 
make[1]: *** [Makefile:1849: /tmp/driver] Error 2 
```

FedeDP commented:
I will document them :) Thanks!

FedeDP commented:
> I would not tie the two things so tight right now...

Mmmh i am not sure: i think we need this; it is the easiest way: the distro builder specifies one of the supported gcc/clang, and based on that, the build processor will choose the correct builder image.
It seems a pretty simple and straightforward logic and would allow us to build a larger number of drivers.

I think i will open a PR once #113  is merged, to achieve that. We can then discuss on it :)

FedeDP commented:
Uh, rebased on top of #113  :1st_place_medal:

FedeDP commented:
/cc @dwindsor :partying_face:

FedeDP commented on pkg/driverbuilder/builder/vanilla.go:
```diff
@@ -91,10 +91,15 @@ func (v vanilla) Script(c Config) (string, error) {
+	kv := kernelReleaseFromBuildConfig(c.Build)
+
+	var urls []string
+	if c.KernelUrls == nil {
```
For me, `kernelurls` is to pass the exact url for downloading a kernel-devel package (in case of Ubuntu, we need 2 urls, the arch one and the `_all` one).

Therefore there is no need to append it.  

Idea is that it is much simpler for an user to search for the exact url for its kernel devel, than to possibly adjust kernels scraping code inside driverkit, thus even if driverkit does not support discovering your url, you can still build your driver.
Moreover, this way we are able to support:
* various ubuntu flavors, like ubuntu-kvm, ubuntu-gcp and so on
* ubuntu/debian kernelversions like `32~16.04`

Note that this is part of a broader scope together with https://github.com/falcosecurity/kernel-crawler, that outputs a driverkit-like json for any found kernel with the urls to download its devel packages; see: https://github.com/falcosecurity/kernel-crawler/blob/main/kernels/x86_64/list.json

FedeDP replied:
What i am trying to say is that in the future users can try their luck by providing exact kernel url parameter whenever a "kernel header not found" error happens; and in case, fixing their configs to provide the `kernelurls`.

In the end, the current kernel url auto-formatting method will remain as a fallback for when no kernelurls param is passed.

That's of course IMHO :)

dwindsor replied:
> Idea is that it is much simpler for an user to search for the exact url for its kernel devel, than to possibly adjust kernels scraping code inside driverkit, thus even if driverkit does not support discovering your url, you can still build your driver.

👆 Definitely. This is a powerful addition!

The case I'm thinking of is, deploying Falco to a fleet of machines that has many instances of "kernel headers not found" because they have many e.g. ubuntu-gcp machines, but with different kernel versions. So, we know the kernel headers for ubuntu-gcp can all be found at e.g. `pub.ubuntu.com/blah/gcp/packages/kernel-devel-1.2.3-4.deb`. Could we have `kernelurls` accept the directory to `pub.ubuntu.com/blah/gcp/packages` and have it crawl that directory via the usual means, auto-extracting the URLs to kernel-devel packages it finds there?

FedeDP replied:
I see; perhaps that would be useful to also use a privare deb repository (for example).

I think it is a great idea; we might think of something like:
* if `kernelurls` contains some formatting options, like `%M.%m.%p` -> we format it with proper variables and use it 
* if `kernelurls` does not contain any formatting option, it means that it is a fullpath and we can use it without further effort

This is a nice solution, even if i'll have to write some formatting specs (like, in the above example: "%M" -> major, "%m" -> minor and "%p" -> patch).

What do you think?
Btw i can also try to split this feature from this PR if you wish; it helped me test various arm64 builds but it is not strictly related to it.

dwindsor replied:
> if `kernelurls` contains some formatting options, like `%M.%m.%p` -> we format it with proper variables and use it

Ahh, I think I see what you mean. So, for e.g. in the previous case mentioned with a fleet of `ubuntu-gcp` machines, one would do something like:

`driverkit --kernelurls pub.ubuntu.com/blah/ubuntu-gcp/packages/kernel-devel-%M.%m-%p.deb`

If so, this is a great idea! Allowing the user to specify the package URL format at runtime protects us against having to constantly update the builders if/when a vendor changes the formatting of their packages.

Doesn't necessarily need to be separate from this PR - I think it makes sense to include it here. We'll get this merged and then cut a new release. 🚀

dwindsor commented:
/lgtm

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/driverkit/pull/143#issuecomment-1142148204" title="LGTM">dwindsor</a>*, *<a href="https://github.com/falcosecurity/driverkit/pull/143#" title="Author self-approved">FedeDP</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fntlnz** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fntlnz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fdriverkit).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/driverkit/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fntlnz"]} -->

FedeDP commented:
/hold

FedeDP commented:
Tested and fixed qemu integration.
Moreover, double tested and fixed `docker buildx` command to build driverkit and builder images.

As agreed with @dwindsor, I will remove the `kernelurls` from this PR because i think it is better to push it in its own one.
Bear with me if i am not able to rebase this PR removing the `kernelurls` and instead add a commit on top to remove it, but i think that it would require manually fixing an subsequent commit tests.
:rocket:

I will unhold the PR once `kernelurls` is removed from this :)

FedeDP commented:
/unhold

Removed `kernelurls` option :)

dwindsor commented:
/lgtm

poiana commented:
LGTM label has been added.  <details>Git tree hash: 49bb73c0e9fae80a82b00376ebfafad8283b45cf</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/driverkit/pull/143#issuecomment-1145876191" title="LGTM">dwindsor</a>*, *<a href="https://github.com/falcosecurity/driverkit/pull/143#" title="Author self-approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fdriverkit).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/driverkit/blob/master/OWNERS)~~ [dwindsor]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/driverkit/pull/143#pullrequestreview-994807592" title="Approved">dwindsor</a>*, *<a href="https://github.com/falcosecurity/driverkit/pull/143#" title="Author self-approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fdriverkit).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/driverkit/blob/master/OWNERS)~~ [dwindsor]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
